library/alloc/src/collections/fixed_queue/mod.rs

```diff
@@ -0,0 +1,288 @@
+pub struct FixedQueue<T, const N: usize> {
+    buffer: [Option<T>; N],
+    head: usize,
+    tail: usize,
+    len: usize,
+}
+
+impl<T, const N: usize> FixedQueue<T, N> {
+    /// Create a new FixedQueue with given fields.
+    #[inline]
+    const fn with(
+        buffer: [Option<T>; N],
+        head: usize,
+        tail: usize,
+        len: usize,
+    ) -> FixedQueue<T, N> {
+        FixedQueue { buffer, head, tail, len }
+    }
+
+    /// Create a new FixedQueue with a given capacity.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub const fn new() -> FixedQueue<T, N>
+    where
+        Option<T>: Copy,
+    {
+        FixedQueue::with([None; N], 0, 0, 0)
+    }
+
+    /// Return the max capacity of the FixedQueue.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub const fn capacity(&self) -> usize {
+        N
+    }
+
+    /// Returns the number of elements in the FixedQueue.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Check if the queue is empty.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub const fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Check if the queue is full.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub const fn is_full(&self) -> bool {
+        self.len == N
+    }
+
+    /// Removes all elements from the queue.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn clear(&mut self) {
+        for i in 0..N {
+            drop(self.buffer[i].take());
+        }
+        self.head = 0;
+        self.tail = 0;
+        self.len = 0;
+    }
+
+    /// Fills the queue with an element.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn fill(&mut self, item: T)
+    where
+        Option<T>: Copy,
+    {
+        self.buffer = [Some(item); N];
+        self.head = 0;
+        self.tail = 0;
+        self.len = N;
+    }
+
+    /// Add an element to the queue. If queue is full, the first element
+    /// is popped and returned.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn push(&mut self, item: T) -> Option<T> {
+        // 'pop' first
+        let overwritten = self.buffer[self.tail].take();
+        // overwrite head/tail element
+        self.buffer[self.tail] = Some(item);
+        // shift tail with 'push'
+        self.tail = (self.tail + 1) % N;
+        if overwritten.is_some() {
+            // shift head ptr on collision
+            self.head = (self.head + 1) % N;
+        } else {
+            // increase len if no collision
+            self.len += 1;
+        }
+        return overwritten;
+    }
+
+    /// Removes and returns the oldest element from the queue.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn pop(&mut self) -> Option<T>
```

Comment:
Like `VecDeque`, this should probably be a `Deque` as well and have `pop_front` and `pop_back`.